plugin/nomad/nomad.go

```diff
@@ -0,0 +1,149 @@
+
+	// Split the query name with a `.` as the delimiter and extract namespace and service name.
+	// If the query is not for a Nomad service, return.
+	qnameSplit := dns.SplitDomainName(qname)
+	if len(qnameSplit) < 3 || qnameSplit[2] != "nomad" {
+		return plugin.NextOrFailure(n.Name(), n.Next, ctx, w, r)
+	}
+	namespace := qnameSplit[1]
+	serviceName := qnameSplit[0]
+
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Authoritative = true
+	m.Compress = true
+	m.Rcode = dns.RcodeSuccess
+	m.Answer = []dns.RR{}
+	m.Extra = []dns.RR{}
+
+	header := dns.RR_Header{
+		Name:   state.QName(),
+		Rrtype: state.QType(),
+		Class:  dns.ClassINET,
+		Ttl:    n.ttl,
+	}
+
+	// Fetch service registrations for the given service.
+	log.Debugf("Looking up record for svc: %s namespace: %s", serviceName, namespace)
+	svcRegistrations, _, err := n.client.Services().Get(serviceName, (&api.QueryOptions{Namespace: namespace}))
+	if err != nil {
+		m.Rcode = dns.RcodeServerFailure
+		w.WriteMsg(m)
+		requestFailedCount.WithLabelValues(metrics.WithServer(ctx), namespace).Inc()
+		return dns.RcodeServerFailure, fmt.Errorf("error fetching service detail: %w", err)
+	}
+
+	// If no service registrations are found, ignore this service.
+	if len(svcRegistrations) == 0 {
+		m.Rcode = dns.RcodeNameError
+		w.WriteMsg(m)
+		requestFailedCount.WithLabelValues(metrics.WithServer(ctx), namespace).Inc()
+		return dns.RcodeNameError, nil
+	}
+
+	// Iterate over all service registrations and add their addresses to the response.
+	for _, s := range svcRegistrations {
+		// Convert address to an IP and add it to the response.
+		addr := net.ParseIP(s.Address)
+		if addr == nil {
+			m.Rcode = dns.RcodeServerFailure
+			w.WriteMsg(m)
+			requestFailedCount.WithLabelValues(metrics.WithServer(ctx), namespace).Inc()
+			return dns.RcodeServerFailure, fmt.Errorf("error parsing IP address: %w", err)
+		}
+
+		// Check the query type to format the appriopriate response.
+		switch qtype {
+		case dns.TypeA:
+			m.Answer = append(m.Answer, &dns.A{
+				Hdr: header,
+				A:   addr,
+			})
+		case dns.TypeAAAA:
+			m.Answer = append(m.Answer, &dns.AAAA{
+				Hdr:  header,
+				AAAA: addr,
+			})
+		case dns.TypeSRV:
+			m.Answer = append(m.Answer, &dns.SRV{
+				Hdr:      header,
+				Target:   qname,
+				Port:     uint16(s.Port),
+				Priority: 10,
+				Weight:   10,
+			})
+			if addr.To4() == nil {
+				m.Extra = append(m.Extra, &dns.AAAA{
+					Hdr: dns.RR_Header{
+						Name:   qname,
+						Rrtype: dns.TypeAAAA,
+						Class:  dns.ClassINET,
+						Ttl:    n.ttl,
+					},
+					AAAA: addr,
+				})
+			} else {
+				m.Extra = append(m.Extra, &dns.A{
+					Hdr: dns.RR_Header{
+						Name:   qname,
+						Rrtype: dns.TypeA,
+						Class:  dns.ClassINET,
+						Ttl:    n.ttl,
+					},
+					A: addr,
+				})
+			}
+		default:
+			m.Rcode = dns.RcodeNotImplemented
+			w.WriteMsg(m)
+			requestFailedCount.WithLabelValues(metrics.WithServer(ctx), namespace).Inc()
+			return dns.RcodeNotImplemented, nil
```

Comment:
I think this will result in sending two responses to the client (the client would completely ignore the second response, so you would not see it).

The first response written above `w.WriteMsg(m)`, the second written by the server when it sees the "failure" response code return value returned from the plugin chain  (dns.RcodeNotImplemented is categorized as a failure by the server). The server uses `ClientWrite` func below to determine if it needs to send an error.

https://github.com/coredns/coredns/blob/e7ad486b50d8f831b7dd4a0840d300a3bcfac471/plugin/plugin.go#L86-L100

Best fix is to remove the `w.WriteMsg(m)` line. 
If you find the response sent by the server is not correct, or needs tweaking, write the response manually as you are doing now, but return dns.RcodeSuccess instead so the server does not send a second response.